datafusion/src/physical_plan/expressions/bitmap_distinct.rs

```diff
@@ -0,0 +1,233 @@
+    fn state(&self) -> Result<Vec<ScalarValue>> {
+        //maybe run optimized
+        let buffer = self.bitmap.serialize();
+        Ok(vec![ScalarValue::Binary(Some(buffer))])
+    }
+
+    fn update_batch(&mut self, values: &[ArrayRef]) -> Result<()> {
+        let value = &values[0];
+        if value.is_empty() {
+            info!("BitmapDistinctCountAccumulator update_batch in empty batch");
+            return Ok(());
+        }
+        match value.data_type() {
+            DataType::Int8 => {
+                let array = value.as_any().downcast_ref::<Int8Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int16 => {
+                let array = value.as_any().downcast_ref::<Int16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::Int32 => {
+                let array = value.as_any().downcast_ref::<Int32Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt8 => {
+                let array = value.as_any().downcast_ref::<UInt8Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt16 => {
+                let array = value.as_any().downcast_ref::<UInt16Array>().unwrap();
+                self.bitmap.add_many(
+                    &array
+                        .values()
+                        .iter()
+                        .map(|&x| x as u32)
+                        .collect::<Vec<u32>>(),
+                );
+            }
+            DataType::UInt32 => {
+                let array = value.as_any().downcast_ref::<UInt32Array>().unwrap();
+                self.bitmap.add_many(array.values());
+            }
+            e => {
+                return Err(DataFusionError::Internal(format!(
+                    "BITMAP_COUNT_DISTINCT is not expected to receive the type {:?}",
+                    e
+                )));
+            }
+        }
+        Ok(())
+    }
+
+    fn merge_batch(&mut self, states: &[ArrayRef]) -> Result<()> {
+        assert_eq!(1, states.len(), "expect only 1 element in the states");
```

Comment:
You have the option to return an error here. Is that perhaps preferable than a hard crash? Maybe higher up the call chain some more context could be added to help debugging.

Reply:
in bitmap_distinct `merge_batch` will get input from `state`, `state` will only return one single array of bitmap, i add it to emphasize it.  
Yes, i should log a error rather than panic👍